gettsim/taxes/zve.py

```diff
@@ -312,58 +316,151 @@ def calc_gross_e7(tax_unit, params):
+    sonstigevors = 12 * (
+        tax_unit["pvbeit"] + (1 - params["vorsorg_krank_minder"]) * tax_unit["gkvbeit"]
     )
-    # These you get anyway ('Basisvorsorge').
-    sonstigevors2010 = 12 * (tax_unit["pvbeit"] + 0.96 * tax_unit["gkvbeit"])
     # maybe add avbeit, but do not exceed 1900€.
-    sonstigevors2010 = np.maximum(
-        sonstigevors2010,
-        np.minimum(
-            sonstigevors2010 + 12 * tax_unit["avbeit"], params["vors_sonst_max"],
-        ),
+    sonstigevors = np.maximum(
+        sonstigevors,
+        np.minimum(sonstigevors + 12 * tax_unit["avbeit"], params["vors_sonst_max"]),
     )
+    return altersvors.astype(int) + sonstigevors.astype(int)
+
+
+def _vorsorge_since_2005(tax_unit, params, soz_vers_beitr_params):
+    """ Vorsorgeaufwendungen 2005 to 2010
+    Pension contributions are accounted for up to €20k.
+    From this, a certain share can actually be deducted,
+    starting with 60% in 2005.
+    Other deductions are simply added up, up to a ceiling of 1500 p.a. for standard employees.
+    """
+
+    altersvors = calc_altersvors_aufwend(tax_unit, params)
+
+    sonstigevors = ~tax_unit["child"] * np.minimum(
+        params["vors_sonst_max"],
+        12 * (tax_unit["gkvbeit"] + tax_unit["avbeit"] + tax_unit["pvbeit"]),
+    ).astype(int)
 
-    vorsorge2010 = altersvors2010 + sonstigevors2010
+    return (altersvors + sonstigevors).astype(int)
 
-    return vorsorge2010.astype(int)
 
+def vorsorge_pre_2005(tax_unit, params, soz_vers_beitr_params):
+    """ Vorsorgeaufwendungen up until 2004.
+        - only pension and health contributions.
+    """
+
+    # Distinguish between married and singles
+    # Single Taxpayer
+    if not tax_unit["zveranl"].max():
+        if params["year"] <= 2019:
+            # Amount 1: Basic deduction, based on earnings. Usually zero.
+            item_1 = np.maximum(
+                params["vorwegab"] - params["kuerzquo"] * 12 * tax_unit["m_wage"], 0
+            )
+        else:
+            # No "vorwegab" anymore after 2019.
+            item_1 = 0
+        # calcuate the remaining amount.
+        vorsorg_rest = np.maximum(
+            12 * (tax_unit["rvbeit"] + tax_unit["gkvbeit"]) - item_1, 0
+        )
+        # Deduct a 'Grundhöchstbetrag' (1334€ in 2004),
+        # or the actual expenses if lower (which is unlikely)
+        item_2 = np.minimum(params["grundbet"], vorsorg_rest)
+        # From what is left from vorsorg_rest, you may deduct 50%.
+        # (up until 50% of 'Grundhöchstbetrag')
+        item_3 = np.minimum(0.5 * (vorsorg_rest - item_2), 0.5 * params["grundbet"])
+    # For the married couple, the same stuff, but with tu totals.
+    if tax_unit["zveranl"].max():
+        for var in ["m_wage", "rvbeit", "gkvbeit"]:
+            tax_unit[f"{var}_tu"] = tax_unit.loc[~tax_unit["child"], "m_wage"].sum()
+        if params["year"] <= 2019:
+            item_1 = 0.5 * np.maximum(
+                2 * params["vorwegab"]
```

Comment:
`vorwegab`doesn't sound very self-explanatory.

Reply:
changed to `vorwegabzug`

Reply:
Same as before. If we abbreviate a word, then mark that with an underscore. I think it is not a good idea to glue together abreviations...

Reply:
But `vorwegabzug` is the whole word, no?

Reply:
Oh sorry. Just googled it. Didn't even know this word. German language is really suprising....
Thought this is some kind of abbreviation.

Reply:
But we didn't include that rule in #60 ?! When I program, I always follow: If I abbreviate something, I mark it with an underscore.

Reply:
"Names follow standard Python conventions (`lowercase_with_underscores`).", but explicit is better than implicit, hence 4feda43.

The whole thing is indeed not fully obvious in German where you just create long words where you would have separate ones in English. 

Anyhow: Good catch!